runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,1963 @@
+	#[pallet::storage]
+	pub(super) type Disputes<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat, SessionIndex,
+		Blake2_128Concat, CandidateHash,
+		DisputeState<T::BlockNumber>,
+	>;
+
+	/// All included blocks on the chain, as well as the block number in this chain that
+	/// should be reverted back to if the candidate is disputed and determined to be invalid.
+	#[pallet::storage]
+	pub(super) type Included<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat, SessionIndex,
+		Blake2_128Concat, CandidateHash,
+		T::BlockNumber,
+	>;
+
+	/// Maps session indices to a vector indicating the number of potentially-spam disputes
+	/// each validator is participating in. Potentially-spam disputes are remote disputes which have
+	/// fewer than `byzantine_threshold + 1` validators.
+	///
+	/// The i'th entry of the vector corresponds to the i'th validator in the session.
+	#[pallet::storage]
+	pub(super) type SpamSlots<T> = StorageMap<_, Twox64Concat, SessionIndex, Vec<u32>>;
+
+	/// Whether the chain is frozen or not. Starts as `false`. When this is `true`,
+	/// the chain will not accept any new parachain blocks for backing or inclusion.
+	/// It can only be set back to `false` by governance intervention.
+	#[pallet::storage]
+	#[pallet::getter(fn is_frozen)]
+	pub(super) type Frozen<T> =  StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub fn deposit_event)]
+	pub enum Event {
+		/// A dispute has been initiated. \[candidate hash, dispute location\]
+		DisputeInitiated(CandidateHash, DisputeLocation),
+		/// A dispute has concluded for or against a candidate.
+		/// \[para_id, candidate hash, dispute result\]
+		DisputeConcluded(CandidateHash, DisputeResult),
+		/// A dispute has timed out due to insufficient participation.
+		/// \[para_id, candidate hash\]
+		DisputeTimedOut(CandidateHash),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Duplicate dispute statement sets provided.
+		DuplicateDisputeStatementSets,
+		/// Ancient dispute statement provided.
+		AncientDisputeStatement,
+		/// Validator index on statement is out of bounds for session.
+		ValidatorIndexOutOfBounds,
+		/// Invalid signature on statement.
+		InvalidSignature,
+		/// Validator vote submitted more than once to dispute.
+		DuplicateStatement,
+		/// Too many spam slots used by some specific validator.
+		PotentialSpam,
+	}
+}
+
+// The maximum number of validators `f` which may safely be faulty.
+//
+// The total number of validators is `n = 3f + e` where `e in { 1, 2, 3 }`.
+fn byzantine_threshold(n: usize) -> usize {
+	n.saturating_sub(1) / 3
+}
+
+// The supermajority threshold of validators which is required to
+// conclude a dispute.
+fn supermajority_threshold(n: usize) -> usize {
+	n - byzantine_threshold(n)
+}
+
+bitflags::bitflags! {
+	#[derive(Default)]
+	struct DisputeStateFlags: u8 {
```

Comment:
Can't this just be an enum instead? All states are mutually exclusive, you can't have a super majority without having a confirmation first.

Reply:
Except when we have `f + 1` equivocators, in which case we might have majority for and against.